cli/admin/src/test/java/io/pravega/cli/admin/dataRecovery/DataRecoveryTest.java

```diff
@@ -0,0 +1,490 @@
+     */
+    private static class ControllerRunner implements AutoCloseable {
+        private final int controllerPort = io.pravega.test.common.TestUtils.getAvailableListenPort();
+        private final String serviceHost = "localhost";
+        private final ControllerWrapper controllerWrapper;
+        private final Controller controller;
+        private final URI controllerURI = URI.create("tcp://" + serviceHost + ":" + controllerPort);
+
+        ControllerRunner(int bkPort, int servicePort, int containerCount) throws InterruptedException {
+            this.controllerWrapper = new ControllerWrapper("localhost:" + bkPort, false,
+                    controllerPort, serviceHost, servicePort, containerCount);
+            this.controllerWrapper.awaitRunning();
+            this.controller = controllerWrapper.getController();
+        }
+
+        @Override
+        public void close() throws Exception {
+            this.controller.close();
+            this.controllerWrapper.close();
+        }
+    }
+
+    /**
+     * Creates a client to read and write events.
+     */
+    private static class ClientRunner implements AutoCloseable {
+        private final ConnectionFactory connectionFactory;
+        private final ClientFactoryImpl clientFactory;
+        private final ReaderGroupManager readerGroupManager;
+
+        ClientRunner(ControllerRunner controllerRunner) {
+            this.connectionFactory = new SocketConnectionFactoryImpl(ClientConfig.builder()
+                    .controllerURI(controllerRunner.controllerURI).build());
+            this.clientFactory = new ClientFactoryImpl(SCOPE, controllerRunner.controller, connectionFactory);
+            this.readerGroupManager = new ReaderGroupManagerImpl(SCOPE, controllerRunner.controller, clientFactory);
+        }
+
+        @Override
+        public void close() {
+            this.readerGroupManager.close();
+            this.clientFactory.close();
+            this.connectionFactory.close();
+        }
+    }
+
+    /**
+     * Creates a Pravega instance.
+     */
+    private static class PravegaRunner implements AutoCloseable {
+        private final int containerCount;
+        private BookKeeperRunner bookKeeperRunner;
+        private SegmentStoreRunner segmentStoreRunner;
+        private ControllerRunner controllerRunner;
+
+        PravegaRunner(int instanceId, int bookieCount, int containerCount, StorageFactory storageFactory) throws Exception {
+            this.containerCount = containerCount;
+            this.bookKeeperRunner = new BookKeeperRunner(instanceId, bookieCount);
+            restartControllerAndSegmentStore(storageFactory, null);
+        }
+
+        public void restartControllerAndSegmentStore(StorageFactory storageFactory, BookKeeperLogFactory dataLogFactory)
+                throws DurableDataLogException, InterruptedException {
+            this.segmentStoreRunner = new SegmentStoreRunner(storageFactory, dataLogFactory, this.containerCount);
```

Comment:
Can this method be invoked multiple times? If not, then enforce it.

If it does, then you need to close whatever segmentStoreRunner and controllerRunner you had in your member variables before assigning them to new instances. Otherwise you'll have resource leaks.

Reply:
Already doing the second thing.